auth/oauth.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2022 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"runtime"
+	"strings"
+
+	atlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+const defaultBaseURL = atlas.CloudURL
+
+var (
+	userAgent = fmt.Sprintf("go-mongodbatlas/%s (%s;%s)", atlas.Version, runtime.GOOS, runtime.GOARCH)
+)
+
+type Config struct {
+	client    *http.Client
+	ClientID  string
+	AuthURL   *url.URL
+	UserAgent string
+	Scopes    []string
+}
+
+type ConfigOpt func(*Config) error
+
+func NewConfig(httpClient *http.Client) *Config {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+
+	baseURL, _ := url.Parse(defaultBaseURL)
+	c := &Config{
+		client:    httpClient,
+		AuthURL:   baseURL,
+		UserAgent: userAgent,
+	}
+	return c
+}
+
+func NewConfigWithOptions(httpClient *http.Client, opts ...ConfigOpt) (*Config, error) {
+	c := NewConfig(httpClient)
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+// SetAuthURL is a config option for setting the base URL.
+func SetAuthURL(bu string) ConfigOpt {
+	return func(c *Config) error {
+		u, err := url.Parse(bu)
+		if err != nil {
+			return err
+		}
+
+		c.AuthURL = u
+		return nil
+	}
+}
+
+// SetUserAgent is a config option for setting the user agent.
+func SetUserAgent(ua string) ConfigOpt {
+	return func(c *Config) error {
+		c.UserAgent = fmt.Sprintf("%s %s", ua, userAgent)
+		return nil
+	}
+}
+
+// SetClientID is a config option for setting the ClientID.
+func SetClientID(clientID string) ConfigOpt {
+	return func(c *Config) error {
+		c.ClientID = clientID
+		return nil
+	}
+}
+
+// SetScopes is a config option for setting the Scopes.
+func SetScopes(scopes []string) ConfigOpt {
+	return func(c *Config) error {
+		c.Scopes = scopes
+		return nil
+	}
+}
+
+// A TokenSource is anything that can return a token.
+type TokenSource interface {
+	// Token returns a token or an error.
+	// Token must be safe for concurrent use by multiple goroutines.
+	// The returned Token must not be modified.
+	Token() (*Token, error)
+}
+
+func (c *Config) Do(ctx context.Context, req *http.Request, v interface{}) (*atlas.Response, error) {
+	resp, err := atlas.DoRequestWithClient(ctx, c.client, req)
+	if err != nil {
+		// If we got an error, and the context has been canceled,
+		// the context's error is probably more useful.
```

Comment:
[q] When this scenario could happen?

Reply:
> If we got an error, and the context has been canceled

https://pkg.go.dev/context

Reply:
Sorry, I didn’t explain well myself. I don’t understand in which scenarios we may get an error without the context.

Reply:
if this scenario happens because the context is cancelled, then we don't need to check the context error if the context has _not_ been cancelled